nilearn/regions/parcellations.py

```diff
@@ -488,24 +553,40 @@ def transform(self, imgs, confounds=None):
         imgs, confounds, single_subject = _check_parameters_transform(
             imgs, confounds
         )
-        # Requires for special cases like extracting signals on list of
-        # 3D images
-        imgs_list = iter_check_niimg(imgs, atleast_4d=True)
-
-        masker = NiftiLabelsMasker(
-            self.labels_img_,
-            mask_img=self.masker_.mask_img_,
-            smoothing_fwhm=self.smoothing_fwhm,
-            standardize=self.standardize,
-            detrend=self.detrend,
-            low_pass=self.low_pass,
-            high_pass=self.high_pass,
-            t_r=self.t_r,
-            resampling_target="data",
-            memory=self.memory,
-            memory_level=self.memory_level,
-            verbose=self.verbose,
-        )
+
+        if isinstance(self.masker_.mask_img_, SurfaceImage):
+            imgs_list = imgs.copy()
+            masker = SurfaceLabelsMasker(
+                self.labels_img_,
+                mask_img=self.masker_.mask_img_,
```

Comment:
So turns out we need SurfaceLabelsMasker as well to work properly here. As far as this PR is concerned, SurfaceLabelsMasker missing two things as of now:
1. doesn't take mask_img to output region signals only for the vertices within the mask.
2. doesn't take confounds that are later used in signal cleaning

Will make separate PRs for each.